### DISCUSSION TITLE
Defining a Quit Key

### DISCUSSION THREAD

**Discussion by jdhitsolutions** [tui-cs/Terminal.Gui#2628] [Q&A] [answered]
I am struggling to define an alternate QuitKey. First off, it seems that if I change the QuitKey in one application, it applies to the next application if I don't specify anything. Is this correct, or am I missing a step after running `Init()`?

I'm pretty sure that setting a single key works,

```powershell
[Terminal.Gui.Application]::QuitKey = [Terminal.Gui.Key]::F12
```

But I can't find a consistent way I've been variations on this in a PowerShell script.

```powershell
[Terminal.Gui.Application]::QuitKey =  ('q' -as [Terminal.Gui.Key])+[Terminal.Gui.Key]::CtrlMask
```

I assume this means I can type Ctrl+q to quit. But it fails. Using Alt+q will work. I even tried using the value.

```powershell
[Terminal.Gui.Application]::QuitKey =1073741937
```

I have to use Alt+q to quit.

I'm running in PowerShell 7 in Windows Terminal. Is this a bug, or is there something I'm not understanding about how to define a quit key? I have not had much luck finding documentation on this.

**Comment by tznind** [comment]
Hi thanks for reaching out.

Application.QuitKey is a static field.  That means changing it in one place changes it permanently across your whole program until you change it again.

But you are free to read and write to the variable as many times as you want.

You can run an application with `Application.Run`.

After the user quits you should call `Application.Shutdown`.  If you want to run another program just call `Application.Init` again then run the next window/view.

I'm afraid I can't help with the powershell syntax but here is how you would do it in c#.

Its important when you want to combine a modifier and a key to use the 'bitwise or' to combine flags.  In C# this is `|`

For example:

```csharp
using myproj;
using Terminal.Gui;

Application.Init();

// Store the default key for later if you want to reset the default
var origQuitKey = Application.QuitKey;

// Use a custom quit key for app 1
Application.QuitKey = Key.AltMask | Key.Q;
var w1 = new Window{Title = "App1 (Alt+Q to quit)"};
Application.Init();
Application.Run(w1);
Application.Shutdown();

// Use a different custom quit key for app 2
Application.QuitKey = origQuit;
var w2 = new Window{Title = "App2 (Ctrl+Q to quit)"};
Application.Init();
Application.Run(w2);
Application.Shutdown();
```

**Comment by jdhitsolutions** [comment]
I am using Init(), Run(), and Shutdown() in all scripts. I can set a single quit key like `Esc` or it seems to work if I set it to use the integer value. I am also getting this error which looks like a .NET error. 
![image](https://github.com/gui-cs/Terminal.Gui/assets/12089920/2b732fff-4a2b-4dfc-b9e3-48929bb1e4c0)
Which might be part of the problem.

**Comment by tznind** [comment]
Interesting maybe there is a problem with how powershell sees classes that have fields differing only in case.

You might be able to avoid the problem with casting from the char.  I don't know how powershell would phrase this but in c# it would be:

```csharp
Application.QuitKey = ((Key)(int)'Q') | Key.AltMask;

Or 


Application.QuitKey = ((Key)(int)'q') | Key.AltMask;
```

**Comment by jdhitsolutions** [comment]
I found the best solution is to define a quit key with an integer value.

**Comment by jdhitsolutions** [comment, marked as answer]
I finally realized the | character was representing a -BOR operation. This all makes more sense to me now.

```powershell
$MenuItem0.ShortCut = [int][Terminal.Gui.Key]"CtrlMask" -bor [int][char]"X"
```

**Reply by BDisp** [reply]
Yes is true it is using a bitwise OR operation.

**Comment by tznind** [comment]
Thanks for taking the time to come back and post when you realized this and
sharing the powershell language for this.  It may help future users!
